### PR TITLE
add retry behavior for nodes

### PR DESCRIPTION
**Pull request by qguyk**
#117 
Attempt for recovery from errors when executing a graph

### CODE REVIEW

File: entropylab/graph_experiment.py [hunk 1]
```diff
@@ -27,11 +28,51 @@
...
+    Output,
+    _NodeExecutionInfo,
+    RetryBehavior,
+)
 from entropylab.instruments.lab_topology import ExperimentResources
 from entropylab.logger import logger
 
 
+def _handle_wait_time(wait_time, backoff, added_delay, maximum_wait_time):
+    wait_time *= backoff
```
Comment (galwiner) [inline entropylab/graph_experiment.py:43]: It seems like the backoff parameter meant to set the time the system waits until the next attempt to run. 
I think the parameter name is not fully indicative of what it does. I wanted to suggest `retry_time` but not even sure that's correct.

File: entropylab/tests/test_retry.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    counter = 1
+    return {"x": 1}
+
+
+def test_retry_behavior(caplog):
+    a1 = PyNode(
+        "a1",
+        a,
+        output_vars={"x"},
+        retry_on_error=RetryBehavior(backoff=2, wait_time=0.2),
```
Comment (galwiner) [inline entropylab/tests/test_retry.py:22]: I would expect the number of max. num of attempts to be set via a parameter to this class but it seems it's set with the node "manifest" - i.e. the function `a()`. I think the counter behaviour is supposed to be part of RetryBehavior

**Review by galwiner — CHANGES_REQUESTED** [review]
I don't know if this is already post the product discussion and the details were agreed, but I think some small tweaking is needed. 
Thanks!

File: entropylab/graph_experiment.py (hunk 1 above)
Comment (qguyk) [reply entropylab/graph_experiment.py:43]: I believe that backoff is a known term https://en.wikipedia.org/wiki/Exponential_backoff

File: entropylab/tests/test_retry.py (hunk 2 above)
Comment (qguyk) [reply entropylab/tests/test_retry.py:22]: every parameter of retry behavior is exposed to the user, it just has default values if you don't specify:
```
@dataclass
class RetryBehavior:
    """
    Attributes:
        number_of_attempts: the maximum number of retries.
        wait_time: the initial wait time between each attempt [seconds].
        backoff: a factor that multiplies the delay on each attempt (1 is no backoff).
        added_delay: delay that is added on each attempt [seconds].
        max_wait_time: optional maximum delay.
    """

    number_of_attempts: float = 5
    wait_time: float = 10
    backoff: float = 2
    added_delay: float = 0
    max_wait_time: Optional[float] = None
```


in the node i just wanted it to fail some time to check that the retry really happens

File: entropylab/graph_experiment.py (hunk 1 above)
Comment (galwiner) [reply entropylab/graph_experiment.py:43]: I didn't know it was, interesting. 
Still, I feel it doesn't tell the user what it does. 
I suppose they can learn,we just need a good docstring

**Review by galwiner — APPROVED** [review]

**Pull request merged**